File: src/lib/src/main/java/work/lclpnet/ap2/impl/game/team/SimpleTeamManager.java
package work.lclpnet.ap2.impl.game.team;

import lombok.Setter;
import net.minecraft.network.chat.MutableComponent;
import net.minecraft.server.level.ServerPlayer;
import net.minecraft.server.players.PlayerList;
import net.minecraft.world.scores.PlayerTeam;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;
import work.lclpnet.ap2.api.game.team.*;
import work.lclpnet.ap2.core.hook.PlayerDisplayNameCallback;
import work.lclpnet.ap2.impl.game.PlayerUtil;
import work.lclpnet.ap2.impl.util.scoreboard.CustomScoreboardManager;
import work.lclpnet.kibu.hook.HookRegistrar;
import work.lclpnet.kibu.hook.player.PlayerConnectionHooks;

import java.util.*;

public class SimpleTeamManager implements TeamManager {

    private final PlayerList playerManager;
    private final TeamConfig teamConfig;
    private final CustomScoreboardManager scoreboard;
    private final PlayerUtil playerUtil;
    private final Map<TeamKey, Team> teams = new HashMap<>();
    private final Map<TeamKey, PlayerTeam> mcTeams = new HashMap<>();
    private final Map<UUID, Team> playerTeams = new HashMap<>();
    private final Set<TeamKey> eliminated = new HashSet<>();
    @Nullable
    private TeamEliminatedListener listener = null;
    @Setter
    private boolean useColorCodes = false;

    public SimpleTeamManager(PlayerList playerManager, TeamConfig teamConfig, CustomScoreboardManager scoreboard, PlayerUtil playerUtil) {
        this.playerManager = playerManager;
        this.teamConfig = teamConfig;
        this.scoreboard = scoreboard;
        this.playerUtil = playerUtil;
    }

    @Override
    public Set<Team> getTeams() {
        return new HashSet<>(teams.values());
    }

    @Override
    public Optional<PlayerTeam> getMinecraftTeam(TeamKey key) {
        return Optional.ofNullable(mcTeams.get(key));
    }

    @Override
    public Optional<Team> getTeam(TeamKey key) {
        return Optional.ofNullable(teams.get(key));
    }

    @Override
    public Optional<Team> getTeam(UUID uuid) {
        return Optional.ofNullable(playerTeams.get(uuid));
    }

    @Override
    public synchronized void partitionIntoTeams(Set<ServerPlayer> players, Set<? extends TeamKey> keys) {
        synchronized (this) {
            reset();

            for (TeamKey key : keys) {
                registerTeam(key);
            }
        }

        // use pre-configured team constellations
        var mapping = teamConfig.getMapping();

        mapping.forEach((player, key) -> {
            Team team = getTeam(key).orElseThrow(() ->
                    new IllegalStateException("Team '%s' is not registered".formatted(key.id())));

            joinTeam(player, team);
        });

        // distribute the rest of the players
        Set<ServerPlayer> notMapped = new HashSet<>(players);
        notMapped.removeAll(mapping.keySet());

        TeamPartitioner partitioner = teamConfig.getPartitioner();
        var partitions = partitioner.splitIntoTeams(notMapped, getTeams());
        partitions.forEach(this::joinTeam);

        playerUtil.updatePlayerListNames(players);
    }

    @Override
    public boolean isParticipating(TeamKey key) {
        return !eliminated.contains(key);
    }

    @Override
    public void setTeamEliminated(Team team) {
        if (!hasTeam(team.key())) return;

        if (eliminated.add(team.key()) && listener != null) {
            listener.teamEliminated(team);
        }
    }

    @Override
    public void bind(TeamEliminatedListener listener) {
        this.listener = listener;
    }

    private boolean hasTeam(TeamKey team) {
        return teams.containsKey(team);
    }

    @Override
    public synchronized Team registerTeam(TeamKey key) {
        SimpleTeam team = createTeam(key);
        teams.put(key, team);
        return team;
    }

    @NotNull
    private SimpleTeam createTeam(TeamKey key) {
        if (hasTeamId(key.id())) {
            throw new IllegalStateException("Duplicate team id '%s'".formatted(key.id()));
        }

        var mcTeam = scoreboard.createTeam(key.id());

        if (useColorCodes) {
            mcTeam.setColor(key.formatting());
        }

        mcTeams.put(key, mcTeam);

        return new SimpleTeam(key, playerManager);
    }

    private boolean hasTeamId(String id) {
        return teams.keySet().stream().anyMatch(key -> key.id().equals(id));
    }

    @Override
    public void joinTeam(ServerPlayer player, Team team) {
        synchronized (this) {
            team.addPlayer(player);
            playerTeams.put(player.getUUID(), team);

            var mcTeam = mcTeams.get(team.key());

            if (mcTeam != null) {
                scoreboard.joinTeam(player, mcTeam);
            }
        }
    }

    private void leaveTeam(ServerPlayer player) {
        synchronized (this) {
            Team team = playerTeams.remove(player.getUUID());

            if (team == null) return;

            team.removePlayer(player);

            var mcTeam = mcTeams.get(team.key());

            if (mcTeam != null) {
                scoreboard.leaveTeam(player, mcTeam);
            }
        }
    }

    private void destroyMcTeam(TeamKey key) {
        scoreboard.removeTeam(key.id());

        synchronized (this) {
            mcTeams.remove(key);
        }
    }

    private synchronized void reset() {
        teams.keySet().forEach(this::destroyMcTeam);
        teams.clear();
        playerTeams.clear();
        mcTeams.clear();
    }

    public void init(HookRegistrar hooks) {
        // move player back into the minecraft team, as they are automatically removed when quitting by the CustomScoreboardManager
        PlayerConnectionHooks.JOIN.registerWith(hooks, player -> {
            PlayerTeam mcTeam;

            synchronized (this) {
                Team team = playerTeams.get(player.getUUID());

                if (team == null) return;

                mcTeam = mcTeams.get(team.key());
            }

            if (mcTeam != null) {
                scoreboard.joinTeam(player, mcTeam);
            }
        });

        PlayerDisplayNameCallback.HOOK.registerWith(hooks, (player, name) -> {
            Team team;

            synchronized (this) {
                team = playerTeams.get(player.getUUID());

                if (team == null) return name;
            }

            MutableComponent text = name instanceof MutableComponent ? (MutableComponent) name : name.copy();

            return text.withStyle(style -> style.withColor(team.key().color()));
        });
    }
}

File: src/lib/src/main/kotlin/work/lclpnet/ap2/game/team/SimpleTeamManager.kt
package work.lclpnet.ap2.game.team

import net.minecraft.network.chat.MutableComponent
import net.minecraft.server.level.ServerPlayer
import net.minecraft.server.players.PlayerList
import net.minecraft.world.scores.PlayerTeam
import work.lclpnet.ap2.core.hook.PlayerDisplayNameCallback
import work.lclpnet.ap2.game.util.PlayerUtil
import work.lclpnet.ap2.impl.game.team.SimpleTeam
import work.lclpnet.ap2.util.scoreboard.CustomScoreboardManager
import work.lclpnet.kibu.hook.HookRegistrar
import work.lclpnet.kibu.hook.player.PlayerConnectionHooks
import java.util.*

class SimpleTeamManager(
    private val playerManager: PlayerList?,
    private val teamConfig: TeamConfig,
    private val scoreboard: CustomScoreboardManager,
    private val playerUtil: PlayerUtil
) : TeamManager {

    val teamsByKey = mutableMapOf<TeamKey, Team>()
    private val mcTeams: MutableMap<TeamKey?, PlayerTeam?> = HashMap<TeamKey?, PlayerTeam?>()
    private val playerTeams: MutableMap<UUID?, Team?> = HashMap<UUID?, Team?>()
    private val eliminated: MutableSet<TeamKey?> = HashSet<TeamKey?>()
    private var listener: TeamEliminatedListener? = null
    private var useColorCodes = false

    override val teams: Set<Team>
        get() = teamsByKey.values.toSet()

    override fun getMinecraftTeam(key: TeamKey): PlayerTeam? =
        mcTeams[key]

    override fun getTeam(key: TeamKey): Team? =
        teamsByKey[key]

    override fun getTeam(uuid: UUID): Team? =
        playerTeams[uuid]

    override fun partitionIntoTeams(players: Iterable<ServerPlayer>, teams: Iterable<TeamKey>) {
        synchronized(this) {
            reset()

            for (key in teams) {
                registerTeam(key)
            }
        }

        // use pre-configured team constellations
        for ((player, key) in teamConfig.mapping) {
            val team = getTeam(key)

            checkNotNull(team) { "Team '${key.id}' is not registered" }

            joinTeam(player, team)
        }

        // distribute the rest of the players
        val notMapped = players.toMutableSet()
        notMapped.removeAll(teamConfig.mapping.keys)

        val partitioner = teamConfig.partitioner
        val partitions = partitioner.splitIntoTeams(notMapped, this.teams)

        for ((player, team) in partitions) {
            joinTeam(player, team)
        }

        playerUtil.updatePlayerListNames(players.toSet())
    }

    override fun isParticipating(key: TeamKey): Boolean {
        return !eliminated.contains(key)
    }

    override fun setTeamEliminated(team: Team) {
        if (!hasTeam(team.key)) return

        if (eliminated.add(team.key)) {
            listener?.teamEliminated(team)
        }
    }

    override fun bind(listener: TeamEliminatedListener?) {
        this.listener = listener
    }

    override fun setUseColorCodes(useColorCodes: Boolean) {
        this.useColorCodes = useColorCodes
    }

    private fun hasTeam(team: TeamKey): Boolean {
        return teamsByKey.containsKey(team)
    }

    @Synchronized
    override fun registerTeam(key: TeamKey): Team {
        val team = createTeam(key)
        teamsByKey[key] = team
        return team
    }

    private fun createTeam(key: TeamKey): SimpleTeam {
        check(!hasTeamId(key.id)) { "Duplicate team id '${key.id}'" }

        val mcTeam = scoreboard.createTeam(key.id)

        if (useColorCodes) {
            mcTeam.color = Optional.of(key.teamColor)
        }

        mcTeams[key] = mcTeam

        return SimpleTeam(key, playerManager)
    }

    private fun hasTeamId(id: String): Boolean =
        teamsByKey.keys.any { key -> key.id == id }

    override fun joinTeam(player: ServerPlayer, team: Team) {
        synchronized(this) {
            team.addPlayer(player)
            playerTeams[player.getUUID()] = team

            val mcTeam = mcTeams[team.key]

            if (mcTeam != null) {
                scoreboard.joinTeam(player, mcTeam)
            }
        }
    }

    private fun leaveTeam(player: ServerPlayer) {
        synchronized(this) {
            val team = playerTeams.remove(player.getUUID()) ?: return

            team.removePlayer(player)

            val mcTeam = mcTeams[team.key]

            if (mcTeam != null) {
                scoreboard.leaveTeam(player, mcTeam)
            }
        }
    }

    private fun destroyMcTeam(key: TeamKey) {
        scoreboard.removeTeam(key.id)

        synchronized(this) {
            mcTeams.remove(key)
        }
    }

    @Synchronized
    private fun reset() {
        for (key in teamsByKey.keys) {
            destroyMcTeam(key)
        }

        teamsByKey.clear()
        playerTeams.clear()
        mcTeams.clear()
    }

    fun init(hooks: HookRegistrar) {
        // move player back into the minecraft team, as they are automatically removed when quitting by the CustomScoreboardManager
        PlayerConnectionHooks.JOIN.registerWith(hooks) { player ->
            val mcTeam = synchronized(this) {
                val team = playerTeams[player.uuid] ?: return@registerWith

                mcTeams[team.key]
            }

            if (mcTeam != null) {
                scoreboard.joinTeam(player, mcTeam)
            }
        }

        PlayerDisplayNameCallback.HOOK.registerWith(hooks) { player, name ->
            val team = synchronized(this) {
                playerTeams[player.uuid]
            } ?: return@registerWith name

            val text = name as? MutableComponent ?: name.copy()

            text.withColor(team.key.color)
        }
    }
}